pkg/tools/etcd_helper.go

```diff
@@ -116,19 +127,70 @@ func (h *EtcdHelper) decodeNodeList(nodes []*etcd.Node, slicePtr interface{}) er
 			}
 			continue
 		}
-		obj := reflect.New(v.Type().Elem())
-		if err := h.Codec.DecodeInto([]byte(node.Value), obj.Interface().(runtime.Object)); err != nil {
-			return err
-		}
-		if h.Versioner != nil {
-			// being unable to set the version does not prevent the object from being extracted
-			_ = h.Versioner.UpdateObject(obj.Interface().(runtime.Object), node)
+		if obj, found := h.getFromCache(node.ModifiedIndex); found {
+			v.Set(reflect.Append(v, reflect.ValueOf(obj)))
+		} else {
+			obj := reflect.New(v.Type().Elem())
+			if err := h.Codec.DecodeInto([]byte(node.Value), obj.Interface().(runtime.Object)); err != nil {
+				return err
+			}
+			if h.Versioner != nil {
+				// being unable to set the version does not prevent the object from being extracted
+				_ = h.Versioner.UpdateObject(obj.Interface().(runtime.Object), node)
+			}
+			v.Set(reflect.Append(v, obj.Elem()))
+			if node.ModifiedIndex != 0 {
+				h.addToCache(node.ModifiedIndex, obj.Interface())
+			}
 		}
-		v.Set(reflect.Append(v, obj.Elem()))
 	}
 	return nil
 }
 
+// etcdCache defines interface used for caching objects stored in etcd. Objects are keyed by
+// their Node.ModifiedIndex, which is unique across all types.
+// All implementations must be thread-safe.
+type etcdCache interface {
+	getFromCache(index uint64) (interface{}, bool)
+	addToCache(index uint64, obj interface{})
+}
+
+const maxEtcdCacheEntries int = 100000
+
+func (h *EtcdHelper) getFromCache(index uint64) (interface{}, bool) {
+	var obj interface{}
```

Comment:
----- Original Message -----
> >  	}
> >  	return nil
> >  }
> >  
> > +// etcdCache defines interface used for caching objects stored in etcd.
> > Objects are keyed by
> > +// their Node.ModifiedIndex, which is unique across all types.
> > +// All implementations must be thread-safe.
> > +type etcdCache interface {
> > +	getFromCache(index uint64) (interface{}, bool)
> > +	addToCache(index uint64, obj interface{})
> > +}
> > +
> > +const maxEtcdCacheEntries int = 100000
> > +
> > +func (h *EtcdHelper) getFromCache(index uint64) (interface{}, bool) {
> > +	var obj interface{}
> 
> Unfortunately we don't have a good metric to measure this, so I will the same
> as I've used initially:
> 
> * e2e performance tests still run ~25% faster (from ~20 min down to ~15 min)
> * cpu usage is slightly higher and profiling shows that DeepCopy uses up to
> 30% (usually around 10%). Still CPU is not saturated and uses 100% only
> rarely (but is very close).

How much more memory?

Fortunately, the work being done for conversion can be used to implement efficient DeepCopy, so I would expect that portion to go down.

> 
> Also latency metrics didn't degrade - 99%tile of api call, except list pods,
> run below 1s (checked manually based on apiserver metrics).
> 
> On a related note - we need a better metric, than e2e total time.

Agree.